### DISCUSSION TITLE
Recommendation functionality (similar to the one in qdrant)

### DISCUSSION THREAD

**Discussion by thomasf1** [milvus-io/milvus#30609] [Q&A and General discussion]
How would I implement a recommendation function similar to [the one in qdrant](https://qdrant.tech/documentation/concepts/explore/) and would adding a dedicated function be something that´s in the plans for milvus?

```python
client.recommend(
    collection_name="{collection_name}",
    positive=[100, 231],
    negative=[718],
    strategy=models.RecommendStrategy.BEST_SCORE,
    limit=10,
)
```

**Comment by xiaofan-luan** [comment]
> How would I implement a recommendation function similar to [the one in qdrant](https://qdrant.tech/documentation/concepts/explore/) and would adding a dedicated function be something that´s in the plans for milvus?
> 
> ```python
> client.recommend(
>     collection_name="{collection_name}",
>     positive=[100, 231],
>     negative=[718],
>     strategy=models.RecommendStrategy.BEST_SCORE,
>     limit=10,
> )
> ```

To make it clearer,  you are searching for vectors closest to positive vectors but far away fro negative vectors?

![Uploading image.png…]()


could your share more about this use cases?

**Reply by xiaofan-luan** [reply]
Or, you are more expecting a filtering functionality.
![Uploading image.png…]()

**Comment by xiaofan-luan** [comment]
For the best score functionality 
"let score = if best_positive_score > best_negative_score {
    best_positive_score;
} else {
    -(best_negative_score * best_negative_score);
};"

**Comment by thomasf1** [comment]
Well, the recommendation function is a bit more complex than that. It seems they also seem to be working on improving the strategies available for recommendations. The details are in the [linked](https://qdrant.tech/documentation/concepts/explore/) documentation.

It basically takes a number of positive and negative ids, gets those vectors and then depending on the strategy gets items that are close to the positive vectors and away from the negative ones (and of course filters out the IDs of the positive/negative examples).

It´s a bit more complex from a standard vector search... Just curious if it´s something you guys are thinking about...

**Comment by xiaofan-luan** [comment]
I've been familiar with this implementation for a while, somewhat resembling ES's "more like this" query.

The question at hand is how to define "close to positive but away from negative" in vector search. Qdrant offers two distinct strategies: average embedding or average score calculation. Average embedding is admittedly a somewhat makeshift solution and struggles to handle negative samples effectively. Average score makes sense to me, though it may not be perfectly suited for application in graph or IVF indexes.

Understanding the use case is crucial, which is why I'm keen to know who is currently utilizing this feature.

Here are some questions to ponder, and feel free to offer your insights:

Regarding positive samples: Do we expect candidates to be close to all samples, or is proximity to any positive sample sufficient?
Concerning negative samples: Should negatives be defined strictly within a certain range, or is being relatively closer to positive than negative enough to qualify as a candidate?
If only negative vectors are provided, do we require AFN (Apporximate Farthest Neighbour) rather than nearest neighbor?
What are the primary use cases for this API?
